File: core/src/main/java/dev/vibris/core/ShaderLink.java
package dev.vibris.core;

interface ShaderLink {
    void switchTo(SourceRegistry.Lease source, OwnershipCheck ownership) throws Failure;

    void detach() throws Failure;

    boolean retainsActiveSource();

    static ShaderLink transientLink() {
        return Transient.INSTANCE;
    }

    @FunctionalInterface
    interface OwnershipCheck {
        void verify() throws Failure;
    }

    final class Failure extends Exception {
        private final boolean stable;

        Failure(String message, boolean stable) {
            super(message);
            this.stable = stable;
        }

        Failure(String message, boolean stable, Throwable cause) {
            super(message, cause);
            this.stable = stable;
        }

        boolean stable() {
            return stable;
        }
    }

    enum Transient implements ShaderLink {
        INSTANCE;

        @Override
        public void switchTo(SourceRegistry.Lease source, OwnershipCheck ownership) throws Failure {
            ownership.verify();
        }

        @Override
        public void detach() {
        }

        @Override
        public boolean retainsActiveSource() {
            return false;
        }
    }
}

File: core/src/main/kotlin/dev/vibris/core/ShaderLink.kt
package dev.vibris.core

internal interface ShaderLink {
    @Throws(Failure::class)
    fun switchTo(source: SourceRegistry.Lease, ownership: OwnershipCheck)

    @Throws(Failure::class)
    fun detach()

    @Throws(Failure::class)
    fun retainsActiveSource(): Boolean

    fun interface OwnershipCheck {
        @Throws(Failure::class)
        fun verify()
    }

    class Failure @JvmOverloads constructor(
        message: String?,
        private val stable: Boolean,
        cause: Throwable? = null,
    ) : Exception(message, cause) {
        fun stable(): Boolean = stable
    }

    enum class Transient : ShaderLink {
        INSTANCE;

        override fun switchTo(source: SourceRegistry.Lease, ownership: OwnershipCheck) {
            ownership.verify()
        }

        override fun detach() = Unit

        override fun retainsActiveSource(): Boolean = false
    }

    companion object {
        @JvmStatic
        fun transientLink(): ShaderLink = Transient.INSTANCE
    }
}